File: basic_dsa_four_hundred_seventy_seven.cpp
/**
 * Basic DSA - Shuffle the Array (LeetCode 1470)
 * nums = [x1,...,xn, y1,...,yn]; return [x1,y1,x2,y2,...,xn,yn].
 */
#include <iostream>
#include <vector>

std::vector<int> shuffle(const std::vector<int>& nums, int n) {
    std::vector<int> out(nums.size());
    for (int i = 0; i < n; ++i) {
        out[2 * i] = nums[i];
        out[2 * i + 1] = nums[n + i];
    }
    return out;
}

void printVec(const std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i)
            std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Shuffle the Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {2, 5, 1, 3, 4, 7};
    printVec(a1);
    std::cout << ", n=3 -> ";
    printVec(shuffle(a1, 3));
    std::cout << "\n";

    std::vector<int> a2 = {1, 2, 3, 4, 4, 3, 2, 1};
    printVec(a2);
    std::cout << ", n=4 -> ";
    printVec(shuffle(a2, 4));
    std::cout << "\n";
}

File: basic_dsa_four_hundred_seventy_seven.java
/**
 * Basic DSA Problems - One Fundamental Array Question
 * Simple and essential DSA problem for beginners
 */
public class basic_dsa_four_hundred_seventy_seven {

    // Problem 1: Majority Element
    // Given an array nums of size n, return the element that appears more than n / 2 times.
    // You may assume the majority element always exists in the input (LeetCode-style).
    // If no majority exists, this implementation returns null after verification.
    //
    // Example:
    // nums = [2,2,1,1,1,2,2] -> 2

    /**
     * Boyer–Moore voting algorithm, then verify frequency
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static Integer majorityElement(int[] nums) {
        if (nums == null || nums.length == 0) {
            return null;
        }

        int candidate = nums[0];
        int count = 0;

        for (int x : nums) {
            if (count == 0) {
                candidate = x;
            }
            count += (x == candidate) ? 1 : -1;
        }

        int freq = 0;
        for (int x : nums) {
            if (x == candidate) {
                freq++;
            }
        }

        return freq > nums.length / 2 ? candidate : null;
    }

    private static String arrayToString(int[] arr) {
        if (arr == null) {
            return "null";
        }
        if (arr.length == 0) {
            return "[]";
        }
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) {
                sb.append(", ");
            }
            sb.append(arr[i]);
        }
        sb.append("]");
        return sb.toString();
    }

    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Majority Element");
        System.out.println("=".repeat(60));

        int[] a1 = {2, 2, 1, 1, 1, 2, 2};
        System.out.println("nums: " + arrayToString(a1));
        System.out.println("Majority: " + majorityElement(a1));

        int[] a2 = {3, 2, 3};
        System.out.println("\nnums: " + arrayToString(a2));
        System.out.println("Majority: " + majorityElement(a2));

        int[] a3 = {1};
        System.out.println("\nnums: " + arrayToString(a3));
        System.out.println("Majority: " + majorityElement(a3));

        int[] a4 = {1, 2, 3};
        System.out.println("\nnums: " + arrayToString(a4));
        System.out.println("Majority: " + majorityElement(a4));

        int[] a5 = {};
        System.out.println("\nnums: " + arrayToString(a5));
        System.out.println("Majority: " + majorityElement(a5));

        System.out.println("\nnums: null");
        System.out.println("Majority: " + majorityElement(null));
    }
}
